server/group/group.js

```diff
@@ -0,0 +1,150 @@
+  try {
+    const user = await User.findByPk(req.session.userId);
+    const newGroup = Group.create({...req.body.group});
+  }
+  catch (error) {
+    next(error);
+  }
+});
+
+router.get('/all/:next', async (req, res, next) => {
+  try {
+    let filters = {};
+    let order = {};
+    if (req.body.filter) {
+      for(let f of req.body.filter) {
+	if(f.hasOwnProperty('zipcode')) {
+	  // All all custom filters here
+	  filters.zipcode = f.zipcode;
+	}
+      }
+      for(let o of req.body.order) {
+	if(o.hasOwnProperty('createdAt')) {
+	  // All all custom order here
+	  filters.createdAt = o.createdAt;
+	}
+      }
+    }
+    const groupsAll = await Group.findAll(
+      [{ offset: req.params.next * 25, limit: 25,
+	 where: { ...filters },
+	 order: [...order],
+       }
+      ]);
+    res.send(groupsAll);
+  }
+  catch (error) {
+    next(error);
+  }
+});
+
+// Gets all users for a specific group
+router.get('/groupusers', async (req, res, next) => {
+  try {
+    const group = Group.findByPk(req.body.groupId);
+    const groupUsers = group.getUsers();
+    res.send(groupUsers);
+  }
+  catch (error) {
+    next(error);
+  }
+});
+
+// Gets all groups for a specific user
+router.get('/user', async (req, res, next) => {
+  try{
+    // I think this is right, I'm going to have to test it.
+    const user = await User.findByPk(req.session.userId);
+    const userGroups = await user.getGroups();
+    res.send(userGroups);
+  }
+  catch (error) {
+    next(error);
+  }
+});
+
+router.post('/create', async (req, res, next) => {
+  try {
+    
+  }
+  catch (error) {
+    next(error);
+  }  
+});
+
+// This will allow a user to add themself to a group, or allow a group owner or admin
+// Add a user to the group of their behalf
+router.post('/addmember', async (req, res, next) => {
+  try{
+    // So we need to check that the logged in user can actually add a person to the group
+    // The client should put checks in place to not let this happen, but who knows what
+    // Sneaky people can do. We need to check on the server as well.
+    let validOwner = GroupMember.findOne({ where: { groupId: req.body.groupId, userId: req.session.userId, isAdmin: true }});
+    if (validOwner) {
+      const group = await GroupMember.create({ userId: req.body.userId, groupId: req.body.groupId});      
+      res.status(201).send();
+    }
+    else {
+      res.status(403).send();
+    }
+  }
+  catch (error) {
+    next(error);
+  }
+});
+
+// This is for an admin/owner to remove a user from a group
+router.delete('/removemember', async (req, res, next) => {
+  try{
+    // So we need to check that the logged in user can actually add a person to the group
+    // The client should put checks in place to not let this happen, but who knows what
+    // Sneaky people can do. We need to check on the server as well.
+    let validOwner = GroupMember.findOne({ where: { groupId: req.body.groupId, userId: req.session.userId, isAdmin: true }});
```

Comment:
Maybe turn this into a piece of middleware?

Reply:
I think that is a poor choice, only based on the fact that group api requests would probably be less or right around 15% of total server requests. I could certainly do that, but up front I was leaving everything contained in its own respective place. 

Let me take a few min and think about that. Because you bring up a valid point @justin-cook-developer

Reply:
You can place middleware in singular routes. It would look like:
`router.get('/me', verifyLoggedIn, async (req, res, next) => {})`

Reply:
what's the difference between putting middleware on a singular route and just writing the validation directly in the route? like what is the advantage?

Reply:
For me personally it just allows for us to be a little my DRY with our code. So we could just apply some piece of logic in the middleware which applies to multiple routes. I think that is really the only advantage. Maybe @justin-cook-developer has other thoughts to advantages.

Reply:
yeah I guess for me it doesn't seem much more DRY if it's only for one route, I get it if it's used for multiple routes

Reply:
Yeah, just DRY and more readable. No need to change that if you don't want to. Just a thought.